File: go/std/bfs_reorder.go
package std

import "sort"

// bfsReorderTables computes the BFS-order permutation tables independently
// of DFS (way B, §9.2): level 0 = roots; for each level L=1..d_max, collect
// nodes at that depth, stable-sort by (new position of parent, rank), and
// assign consecutive positions. Mirrors
// cxx/traaxx/sources/bfs_reorder.cpp's bfs_reorder_tables.
func bfsReorderTables(parent, sibling []Index) (ReorderTables, error) {
	n := len(parent)
	if n == 0 {
		return ReorderTables{NewToOld: []Index{}, OldToNew: []Index{}}, nil
	}
	nodeDepth, err := Depth(parent, Index(n))
	if err != nil {
		return ReorderTables{}, err
	}
	nodeRank, err := Rank(sibling, Index(n))
	if err != nil {
		return ReorderTables{}, err
	}
	dMax := Index(0)
	for _, d := range nodeDepth {
		if d > dMax {
			dMax = d
		}
	}
	new2old := make([]Index, n)
	old2new := make([]Index, n)
	position := Index(0)
	for level := Index(0); level <= dMax; level++ {
		levelNodes := []Index{}
		for i := 0; i < n; i++ {
			if nodeDepth[i] == level {
				levelNodes = append(levelNodes, Index(i))
			}
		}
		sort.SliceStable(levelNodes, func(a, b int) bool {
			nodeA, nodeB := levelNodes[a], levelNodes[b]
			var keyA, keyB Index
			if level != 0 {
				keyA = old2new[parent[nodeA]]
				keyB = old2new[parent[nodeB]]
			}
			if keyA != keyB {
				return keyA < keyB
			}
			return nodeRank[nodeA] < nodeRank[nodeB]
		})
		for _, oldID := range levelNodes {
			new2old[position] = oldID
			old2new[oldID] = position
			position++
		}
	}
	return ReorderTables{NewToOld: new2old, OldToNew: old2new}, nil
}

// BfsReorderResult is the read-only, copy-returning form's result. Mirrors
// traaxx::BfsReorderResult<T,IndexT>.
type BfsReorderResult[T any] struct {
	Tree     *Tree[T]
	NewToOld []Index
	OldToNew []Index
}

// BfsReorder computes the BFS reorder of tree without mutating it. Mirrors
// the free-function form traaxx::bfs_reorder(const Tree<T,IndexT>&).
func BfsReorder[T any](tree *Tree[T]) (BfsReorderResult[T], error) {
	tables, err := bfsReorderTables(tree.Parent(), tree.Sibling())
	if err != nil {
		return BfsReorderResult[T]{}, err
	}
	reordered := NewTree(
		RemapIndexVector(tree.Parent(), tables.NewToOld, tables.OldToNew),
		RemapIndexVector(tree.Sibling(), tables.NewToOld, tables.OldToNew),
		Gather(tree.Data(), tables.NewToOld),
	)
	return BfsReorderResult[T]{Tree: reordered, NewToOld: tables.NewToOld, OldToNew: tables.OldToNew}, nil
}

// BfsReorder mutates the tree in place into BFS order, returning only the
// permutation tables. Mirrors the method form
// traaxx::Tree<T,IndexT>::bfs_reorder().
func (t *Tree[T]) BfsReorder() (ReorderTables, error) {
	tables, err := bfsReorderTables(t.parent, t.sibling)
	if err != nil {
		return ReorderTables{}, err
	}
	newParent := RemapIndexVector(t.parent, tables.NewToOld, tables.OldToNew)
	newSibling := RemapIndexVector(t.sibling, tables.NewToOld, tables.OldToNew)
	newData := Gather(t.data, tables.NewToOld)
	t.parent = newParent
	t.sibling = newSibling
	t.data = newData
	return tables, nil
}

File: cxx/traaxx_simd/sources/bfs_reorder.cpp
#include <traaxx_simd/bfs_reorder.hpp>
#include <algorithm>
#include <traaxx_simd/depth.hpp>
#include <traaxx_simd/rank.hpp>

namespace traaxx_simd
{
    // Per-level std::stable_sort — control-flow-heavy, level sizes are
    // data-dependent, not a batch::gather/select shape; matches cxx/traaxx.
    template<typename IndexT>
    std::expected<ReorderTables<IndexT>, ConvergenceError<IndexT>> bfs_reorder_tables(
        const std::vector<IndexT> &parent, const std::vector<IndexT> &sibling)
    {
        auto const n = parent.size();
        if (n == 0)
        {
            return ReorderTables<IndexT>{};
        }
        auto depth_result = depth(parent, static_cast<IndexT>(n));
        if (!depth_result.has_value())
        {
            return std::unexpected(depth_result.error());
        }
        auto rank_result = rank(sibling, static_cast<IndexT>(n));
        if (!rank_result.has_value())
        {
            return std::unexpected(rank_result.error());
        }
        auto const &node_depth = *depth_result;
        auto const &node_rank = *rank_result;
        auto const d_max = *std::max_element(node_depth.cbegin(), node_depth.cend());
        auto new2old = std::vector<IndexT>(n);
        auto old2new = std::vector<IndexT>(n, IndexT{ 0 });
        auto position = IndexT{ 0 };
        for (IndexT level = 0; level <= d_max; ++level)
        {
            auto level_nodes = std::vector<IndexT>{};
            for (IndexT i = 0; i < static_cast<IndexT>(n); ++i)
            {
                if (node_depth[i] == level)
                {
                    level_nodes.push_back(i);
                }
            }
            std::stable_sort(level_nodes.begin(), level_nodes.end(),
                [&](IndexT a, IndexT b)
                {
                    auto const key_a = level == IndexT{ 0 } ? IndexT{ 0 } : old2new[parent[a]];
                    auto const key_b = level == IndexT{ 0 } ? IndexT{ 0 } : old2new[parent[b]];
                    if (key_a != key_b)
                    {
                        return key_a < key_b;
                    }
                    return node_rank[a] < node_rank[b];
                });
            for (auto const old_id : level_nodes)
            {
                new2old[position] = old_id;
                old2new[old_id] = position;
                ++position;
            }
        }
        return ReorderTables<IndexT>{ std::move(new2old), std::move(old2new) };
    }

    template<typename T, typename IndexT>
    std::expected<ReorderTables<IndexT>, ConvergenceError<IndexT>> Tree<T, IndexT>::bfs_reorder()
    {
        auto tables = bfs_reorder_tables(parent_, sibling_);
        if (!tables.has_value())
        {
            return std::unexpected(tables.error());
        }
        auto new_parent = remap_index_vector(parent_, tables->new_to_old, tables->old_to_new);
        auto new_sibling = remap_index_vector(sibling_, tables->new_to_old, tables->old_to_new);
        auto new_data = gather(data_, tables->new_to_old);
        parent_.swap(new_parent);
        sibling_.swap(new_sibling);
        data_.swap(new_data);
        return std::move(tables.value());
    }

    template std::expected<ReorderTables<std::uint32_t>, ConvergenceError<std::uint32_t>> bfs_reorder_tables<
        std::uint32_t>(const std::vector<std::uint32_t> &, const std::vector<std::uint32_t> &);

    template std::expected<ReorderTables<std::uint32_t>, ConvergenceError<std::uint32_t>>
    Tree<std::uint32_t, std::uint32_t>::bfs_reorder();
}
